### PR TITLE
Propagate EC throughout codebase

### PR DESCRIPTION
**Pull request by rohan-sircar**
Earlier EC was hardcoded at some places, now it is propagated
as an implicit parameter throughout the codebase.
While a best practice by itself, it should also hopefully fix
issues with traces and MDC disappearing.

Actual EC is injected only once in  "end of the world" spots
like main method or test cases, and DiagnosticContextFuturePropagation
is used, with a `global` value that wraps around scala's global EC.

### CODE REVIEW

File: modules/common/src/main/scala/surge/internal/akka/cluster/Shard.scala [hunk 1]
```diff
@@ -35,13 +35,15 @@ object Shard {
...
+      implicit tracer: Tracer,
+      ec: ExecutionContext): Props = {
+    Props(new Shard(shardId, regionLogicProvider, extractEntityId)(tracer, ec))
   }
 }
 
 class Shard[IdType](shardId: String, regionLogicProvider: PerShardLogicProvider[IdType], extractEntityId: PartialFunction[Any, IdType])(
-    implicit val tracer: Tracer)
+    implicit val tracer: Tracer,
+    ec: ExecutionContext)
```
Comment (jeffboutotte) [inline modules/common/src/main/scala/surge/internal/akka/cluster/Shard.scala:46]: Does this actor actually use an ExecutionContext anywhere? In general my preference for the EC to use within an actor is the actors dispatcher. As you performance tune or separate out actors to their own dispatcher it makes it much easier to reason about which actors are using what and ensuring things are isolated at that level.

File: modules/command-engine/javadsl/src/main/scala/surge/javadsl/command/SurgeCommand.scala [hunk 2]
```diff
@@ -32,15 +33,14 @@ trait SurgeCommand[AggId, Agg, Command, Evt] extends core.SurgeProcessingTrait[A
 }
 
 object SurgeCommand {
-  def create[AggId, Agg, Command, Evt](businessLogic: SurgeCommandBusinessLogicTrait[AggId, Agg, Command, Evt]): SurgeCommand[AggId, Agg, Command, Evt] = {
+  def create[AggId, Agg, Command, Evt](businessLogic: SurgeCommandBusinessLogicTrait[AggId, Agg, Command, Evt])(
```
[the remark this answers is not in the document]
Comment (rohan-sircar) [reply modules/command-engine/javadsl/src/main/scala/surge/javadsl/command/SurgeCommand.scala:36]: You are right, that's an oversight on my part. I thought I had taken care of it..

File: modules/command-engine/javadsl/src/main/scala/surge/javadsl/event/SurgeEvent.scala [hunk 3]
```diff
@@ -26,17 +27,17 @@ trait SurgeEvent[AggId, Agg, Evt] extends core.SurgeProcessingTrait[Agg, Nothing
 }
 
 object SurgeEvent {
-  def create[AggId, Agg, Evt](businessLogic: SurgeEventBusinessLogic[AggId, Agg, Evt]): SurgeEvent[AggId, Agg, Evt] = {
+  def create[AggId, Agg, Evt](businessLogic: SurgeEventBusinessLogic[AggId, Agg, Evt], ec: ExecutionContext): SurgeEvent[AggId, Agg, Evt] = {
```
Comment (rohan-sircar) [inline modules/command-engine/javadsl/src/main/scala/surge/javadsl/event/SurgeEvent.scala:30]: @jeffboutotte taking it as a regular parameter in the javadsl like here okay right?

File: modules/common/src/main/scala/surge/internal/utils/ExecutionContextProber.scala [hunk 4]
```diff
@@ -78,7 +78,6 @@ class ExecutionContextProberActor(settings: ExecutionContextProberSettings) exte
   def ready(detectedIssue: Boolean): Receive = {
     case SendProbes =>
       val id = UUID.randomUUID()
-      implicit val ec: ExecutionContext = settings.targetEc
```
[the remark this answers is not in the document]
Comment (rohan-sircar) [reply modules/common/src/main/scala/surge/internal/utils/ExecutionContextProber.scala:81]: I'll revert this change

File: modules/common/src/main/scala/surge/internal/akka/cluster/Shard.scala (hunk 1 above)
Comment (rohan-sircar) [reply modules/common/src/main/scala/surge/internal/akka/cluster/Shard.scala:46]: Makes sense, I'll look into removing the ec param.

File: modules/surge-docs/src/test/scala/docs/command/BankAccountEngine.scala [hunk 5]
```diff
@@ -2,12 +2,14 @@
...
 
-import java.util.UUID
+import surge.internal.utils.DiagnosticContextFuturePropagation
 
+import java.util.UUID
 import surge.scaladsl.command.SurgeCommand
 
 // #bank_account_engine_class
 object BankAccountEngine {
+  implicit val ec = DiagnosticContextFuturePropagation.global
```
[the remark this answers is not in the document]
Comment (rohan-sircar) [reply modules/surge-docs/src/test/scala/docs/command/BankAccountEngine.scala:12]: We could do it internally, we'll just need some code to check if the passed EC is already a DCFP EC.

**Review by jeffboutotte — APPROVED** [review]
Awesome - thanks!

**Pull request merged**